File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    int lowerBound(int[] nums, int target) {
        int l = 0;
        int r = nums.length-1;
        int res = nums.length;
        while(l <= r) {
            int mid = l + (r-l)/2;
            if(nums[mid] >= target) {
                res = mid;
                r = mid - 1;
            } else l = mid + 1;
        }
        return res;
    }

    int upperBound(int[] nums, int target) {
        int l = 0;
        int r = nums.length-1;
        int res = nums.length;
        while(l <= r) {
            int mid = l + (r-l)/2;
            if(nums[mid] > target) {
                res = mid;
                r = mid - 1;
            } else l = mid + 1;
        }
        return res;
    }

    public int[] searchRange(int[] nums, int target) {
        if(nums.length == 0) return new int[] {-1, -1};
        int x = lowerBound(nums, target);
        if(x == nums.length || nums[x] != target) return new int[] {-1, -1};
        return new int[] {x, upperBound(nums, target) -1};
    }
}

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def lower_bound(self, nums, target):
        n = len(nums)
        l, r = 0, n - 1
        ans = n

        while l <= r:
            mid = (l + r) // 2

            if nums[mid] >= target:
                ans = mid
                r = mid - 1
            else:
                l = mid + 1

        return ans

    def upper_bound(self, nums, target):
        n = len(nums)
        l, r = 0, n - 1
        ans = n

        while l <= r:
            mid = (l + r) // 2

            if nums[mid] > target:
                ans = mid
                r = mid - 1
            else:
                l = mid + 1

        return ans

    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left = self.lower_bound(nums, target)

        if left == len(nums) or nums[left] != target:
            return [-1, -1]

        right = self.upper_bound(nums, target) - 1

        return [left, right]